elastic/kibana
# [Response Ops][Alerting] Retry installation of alerts as data resources with each rule execution

Resolves https://github.com/elastic/kibana/issues/159038

## Summary

At the start of rule execution, the alerting task runner checks whether all the resources for the current rule type have been installed. If there was an error during installation, either of the common resources (ILM policy, component templates) or the rule type specific resources (index templates, concrete write index), then resource installation will be retried. 

This PR also removes the logic that turns off rule registry writing during the first error it encounters. Previously, if resource installation failed, the rule registry would disable writing for the rule type during the first execution. This meant that even if the framework level retry succeeded on the second rule execution, the rule registry would still not write any alerts. Removing this logic will enable the rule registry to start writing when and if a retry of the resource installation succeeds.

There will be a followup issue to create jest integration tests for this.

## To Verify

**1. Test that if one of the common resources fail to install on startup, installation is retried on each execution and the rule eventually recovers**
* Start ES and Kibana
* Create a metric threshold rule and a detection rule and let them run
* Make the following change to `x-pack/plugins/alerting/server/alerts_service/alerts_service.ts` and restart Kibana

```
--- a/x-pack/plugins/alerting/server/alerts_service/alerts_service.ts
+++ b/x-pack/plugins/alerting/server/alerts_service/alerts_service.ts
@@ -112,7 +112,7 @@ export class AlertsService implements IAlertsService {
   private resourceInitializationHelper: ResourceInstallationHelper;
   private registeredContexts: Map<string, IRuleTypeAlerts> = new Map();
   private commonInitPromise: Promise<InitializationPromise>;
-
+  private retries: number = 0;
   constructor(private readonly options: AlertsServiceParams) {
     this.initialized = false;

@@ -287,7 +287,7 @@ export class AlertsService implements IAlertsService {
           createOrUpdateIlmPolicy({
             logger: this.options.logger,
             esClient,
-            name: DEFAULT_ALERTS_ILM_POLICY_NAME,
+            name: this.retries++ > 10 ? DEFAULT_ALERTS_ILM_POLICY_NAME : '',
             policy: DEFAULT_ALERTS_ILM_POLICY,
           }),
         () =>
```
This will force an error when installing the ILM policy (a common resource) the first 10 times but succeed the 11th time. You will see error messages in the logs each time the retry fails
* Verify that the metric threshold rule and the detection rule fail for a few executions until the 11th retry but then execution succeeds and alerts get written to their respective indices. Note that since this is a common resource, installation will get retried each time either of the rule runs, not each.
* After the 11th retry, you should see the following in the logs:
```
[2023-06-21T11:48:17.980-04:00][INFO ][plugins.alerting] Retrying common resource initialization
[2023-06-21T11:48:17.980-04:00][INFO ][plugins.alerting] Retrying resource initialization for context "observability.metrics"
[2023-06-21T11:48:17.980-04:00][INFO ][plugins.alerting] Installing ILM policy .alerts-ilm-policy
[2023-06-21T11:48:17.981-04:00][INFO ][plugins.alerting] Installing component template .alerts-framework-mappings
[2023-06-21T11:48:17.981-04:00][INFO ][plugins.alerting] Installing component template .alerts-legacy-alert-mappings
[2023-06-21T11:48:17.992-04:00][INFO ][plugins.alerting] Installing component template .alerts-ecs-mappings
[2023-06-21T11:48:17.996-04:00][INFO ][plugins.alerting] Retrying common resource initialization
[2023-06-21T11:48:17.996-04:00][INFO ][plugins.alerting] Retrying resource initialization for context "security"
[2023-06-21T11:48:17.997-04:00][INFO ][plugins.alerting] Installing ILM policy .alerts-ilm-policy
```
and the rules should start running successfully and writing alerts


**2. Test that if one of the context-specific resources fail to install on startup, installation is retried on each execution and the rule eventually recovers** 
* Start ES and Kibana
* Create a metric threshold rule and a detection rule and let them run
* Make the following change to `x-pack/plugins/alerting/server/alerts_service/alerts_service.ts` and restart Kibana

```
--- a/x-pack/plugins/alerting/server/alerts_service/alerts_service.ts
+++ b/x-pack/plugins/alerting/server/alerts_service/alerts_service.ts
@@ -112,7 +112,7 @@ export class AlertsService implements IAlertsService {
   private resourceInitializationHelper: ResourceInstallationHelper;
   private registeredContexts: Map<string, IRuleTypeAlerts> = new Map();
   private commonInitPromise: Promise<InitializationPromise>;
-
+  private retries: number = 0;
   constructor(private readonly options: AlertsServiceParams) {
     this.initialized = false;

@@ -400,6 +400,10 @@ export class AlertsService implements IAlertsService {
     componentTemplateRefs.push(getComponentTemplateName());

     // Context specific initialization installs index template and write index
+    let totalFieldsLimit = TOTAL_FIELDS_LIMIT;
+    if (context === 'observability.metrics') {
+      totalFieldsLimit = this.retries++ > 10 ? TOTAL_FIELDS_LIMIT : 3;
+    }
     initFns = initFns.concat([
       async () =>
         await createOrUpdateIndexTemplate({
@@ -411,14 +415,14 @@ export class AlertsService implements IAlertsService {
             indexPatterns: indexTemplateAndPattern,
             kibanaVersion: this.options.kibanaVersion,
             namespace,
-            totalFieldsLimit: TOTAL_FIELDS_LIMIT,
+            totalFieldsLimit,
           }),
         }),
       async () =>
         await createConcreteWriteIndex({
           logger: this.options.logger,
           esClient,
-          totalFieldsLimit: TOTAL_FIELDS_LIMIT,
+          totalFieldsLimit,
           indexPatterns: indexTemplateAndPattern,
         }),
     ]);
```
This will force an error when installing resources for `observability.metrics` which is used by the metric threshold rule the first 10 times but succeed the 11th time. You will see error messages in the logs each time the retry fails
* Verify that the detection rule (different context) continues to run successfully
* Verify that the metric threshold rule fail for a few executions until the 11th retry but then execution succeeds and alerts get. 
* After the 11th retry, you should see the following in the logs:
```
[2023-06-21T12:17:15.202-04:00][INFO ][plugins.alerting] Retrying resource initialization for context "observability.metrics"
[2023-06-21T12:17:15.202-04:00][INFO ][plugins.alerting] Installing component template .alerts-observability.metrics.alerts-mappings
[2023-06-21T12:17:15.205-04:00][INFO ][plugins.alerting] Installing index template .alerts-observability.metrics.alerts-default-index-template
[2023-06-21T12:17:15.233-04:00][INFO ][plugins.alerting] Creating concrete write index - .internal.alerts-observability.metrics.alerts-default-000001
[2023-06-21T12:17:15.324-04:00][INFO ][plugins.alerting] Resource installation for "observability.metrics" succeeded after retry
```
and the rules should start running successfully and writing alerts
